src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/CallConverterThunk.cs

```diff
@@ -1128,15 +1129,36 @@ unsafe private static void InvokeTarget(void* allocatedStackBuffer, ref CallConv
                         if (!conversionParams._callerArgs.HasRetBuffArg() && conversionParams._calleeArgs.HasRetBuffArg())
                             returnValueToCopy = (void*)(new IntPtr(*((void**)returnValueToCopy)) + IntPtr.Size);
 
+                        if (returnType == CorElementType.ELEMENT_TYPE_BYREF)
+                        {
+                            // If this is a byref return, we're going to dereference the result
+                            returnValueToCopy = *(void**)returnValueToCopy;
+                        }
+
+                        RuntimeTypeHandle returnTypeRuntimeTypeHandle = thRetType.GetRuntimeTypeHandle();
+
                         // Need to box value type before returning it
                         object returnValue;
-                        if (thRetType.IsPointerType())
+                        if (returnType == CorElementType.ELEMENT_TYPE_BYREF && returnValueToCopy == null)
+                        {
+                            // This is a byref return and dereferencing it would result in a NullReferenceException.
+                            // Set the return value to a sentinel that InvokeUtils will recognize.
+                            // Can't throw from here or we would wrap this in a TargetInvocationException.
+                            returnValue = InvokeUtils.NullByRefValueSentinel;
+                        }
+                        else if (RuntimeAugments.IsUnmanagedPointerType(returnTypeRuntimeTypeHandle))
+                        {
+                            returnValue = System.Reflection.Pointer.Box(*(void**)returnValueToCopy, Type.GetTypeFromHandle(returnTypeRuntimeTypeHandle));
+                        }
+                        else if (RuntimeAugments.IsValueType(returnTypeRuntimeTypeHandle))
                         {
-                            returnValue = System.Reflection.Pointer.Box(*(void**)returnValueToCopy, Type.GetTypeFromHandle(thRetType.GetRuntimeTypeHandle()));
+                            returnValue = RuntimeAugments.Box(returnTypeRuntimeTypeHandle, new IntPtr(returnValueToCopy));
                         }
                         else
                         {
-                            returnValue = RuntimeAugments.Box(thRetType.GetRuntimeTypeHandle(), new IntPtr(returnValueToCopy));
+                            // byref return of a reference type
+                            Debug.Assert(returnType == CorElementType.ELEMENT_TYPE_BYREF);
+                            returnValue = Unsafe.As<byte, object>(ref *(byte*)returnValueToCopy);
```

Comment:
Nit: Can you move this with the block a few lines above, to have a single block for if (returnType == CoreElementType.ELEMENT_TYPE_BYREF)

Reply:
I think it needs to stay split this way. The complexity is necessary to e.g. handle the cases of `ref int*` where if the ref is null, we want this to return the `NullByRefValueSentinel`, but if it's non-null, we need it to take the `Pointer.Box` path to box the dereferenced value.

Reply:
I see. Good point, and good test that can be added

Reply:
Yeah, I have a boatload of tests on the side for these. Ended up finding [CLR bugs](https://github.com/dotnet/coreclr/issues/19307) too. I'll add them to the review when I move this to the TFS side.